src/hashtable.c

```diff
@@ -0,0 +1,2135 @@
+    }
+}
+
+/* Clears the position of the entry in the hashtable and resumes rehashing. The
+ * entry destructor is NOT called. The position is acquired using a preceding
+ * call to hashtableTwoPhasePopFindRef(). */
+void hashtableTwoPhasePopDelete(hashtable *ht, hashtablePosition *position) {
+    /* Read position. */
+    struct position *p = positionFromOpaque(position);
+    bucket *b = p->bucket;
+    int pos_in_bucket = p->pos_in_bucket;
+    int table_index = p->table_index;
+
+    /* Delete the entry and resume rehashing. */
+    assert(isPositionFilled(b, pos_in_bucket));
+    b->presence &= ~(1 << pos_in_bucket);
+    ht->used[table_index]--;
+    hashtableShrinkIfNeeded(ht);
+    hashtableResumeRehashing(ht);
+    if (b->chained && !hashtableIsRehashingPaused(ht)) {
+        /* Rehashing paused also means bucket chain compaction paused. It is
+         * paused while iterating and when a scan callback is running, to be
+         * able to live up to the scan and iterator guarantees. In those cases,
+         * we do the compaction in the scan and iterator code instead. */
+        fillBucketHole(ht, b, pos_in_bucket, table_index);
+    }
+}
+
+/* Initializes the state for an incremental find operation.
+ *
+ * Incremental find can be used to speed up the loading of multiple objects by
+ * utilizing CPU branch predictions to parallelize memory accesses. Initialize
+ * the data for a number of incremental find operations. Then call
+ * hashtableIncrementalFindStep on them in a round-robin order until all of them
+ * are complete. Finally, if necessary, call hashtableIncrementalFindGetResult.
+ */
+void hashtableIncrementalFindInit(hashtableIncrementalFindState *state, hashtable *ht, const void *key) {
+    incrementalFind *data = incrementalFindFromOpaque(state);
+    if (hashtableSize(ht) == 0) {
+        data->state = HASHTABLE_NOT_FOUND;
+    } else {
+        data->state = HASHTABLE_NEXT_BUCKET;
+        data->bucket = NULL;
+        data->hashtable = ht;
+        data->key = key;
+        data->hash = hashKey(ht, key);
+    }
+}
+
+/* Returns 1 if more work is needed, 0 when the done. Call this function
+ * repeatedly until it returns 0. Then use hashtableIncrementalFindGetResult to
+ * fetch the result. */
+int hashtableIncrementalFindStep(hashtableIncrementalFindState *state) {
+    incrementalFind *data = incrementalFindFromOpaque(state);
+    switch (data->state) {
+    case HASHTABLE_CHECK_ENTRY:
+        /* Current entry is prefetched. Now check if it's a match. */
+        {
+            hashtable *ht = data->hashtable;
+            void *entry = data->bucket->entries[data->pos];
+            const void *elem_key = entryGetKey(ht, entry);
+            /* printf("Checking entry at pos %d for key=%p vs elem=%p\n", */
+            /*        data->pos, data->key, elem_key); */
+            if (compareKeys(ht, data->key, elem_key) == 0) {
+                /* It's a match. */
+                data->state = HASHTABLE_FOUND;
```

Comment:
Shouldn't we prefetch the value here if it exists and is not embedded?

Reply:
The current entry is already prefetched here. The hashtable abstraction itself can't do more. In the calling code (memory_prefetch.c) you can call hashtableIncrementalFindGetResult and prefetch `robj->ptr`?